File: docs/secure-messaging-architecture/client/ios/SecureMessengerDemo.swiftpm/Sources/AppModel.swift
//
//  AppModel.swift
//
//  Verbindet Identität, Kontakt, Transport und Krypto zu dem in
//  ../../../PROTOCOL.md beschriebenen Ablauf.
//

import Foundation

struct ChatMessage: Identifiable {
    let id = UUID()
    let text: String
    let mine: Bool
}

struct LogEntry: Identifiable {
    let id = UUID()
    let text: String
    let kind: LogKind
}

@MainActor
final class AppModel: ObservableObject {
    @Published private(set) var me: Identity
    @Published var contact: ContactIdentity?
    @Published var contactInputText: String = ""
    @Published var sendText: String = ""
    @Published private(set) var messages: [ChatMessage] = []
    @Published private(set) var techLog: [LogEntry] = []

    @Published var connectionStatus: ConnectionStatus = .connecting
    @Published var connectionStatusText: String = "Verbinde…"

    let crypto = DemoCrypto()
    let transport = Transport()

    private let contactKey = "cipherwire_contact_v1"
    private let historyKey = "cipherwire_history_v1"

    init() {
        me = IdentityStore.loadOrCreate(crypto: crypto)
        loadContact()
        loadHistory()

        transport.onLog = { [weak self] text, kind in
            Task { @MainActor in self?.log(text, kind) }
        }
        transport.onEnvelope = { [weak self] envelope in
            Task { @MainActor in self?.handleIncoming(envelope) }
        }
        transport.$status.receive(on: DispatchQueue.main).assign(to: &$connectionStatus)
        transport.$statusText.receive(on: DispatchQueue.main).assign(to: &$connectionStatusText)

        transport.connect(ownBoxPubHex: me.boxPublicKey.hexString)
        log("Meine ID: " + me.idString, .info)
    }

    func log(_ text: String, _ kind: LogKind) {
        techLog.append(LogEntry(text: text, kind: kind))
    }

    func addContact() {
        guard let parsed = crypto.parseContactIdString(contactInputText) else {
            log("✗ Ungültiges ID-Format (erwartet: 64 Hex . 64 Hex)", .fail)
            return
        }
        contact = parsed
        messages = []
        contactInputText = ""
        UserDefaults.standard.set(
            parsed.boxPublicKey.hexString + "." + parsed.signPublicKey.hexString,
            forKey: contactKey
        )
        saveHistory()
    }

    func removeContact() {
        contact = nil
        messages = []
        UserDefaults.standard.removeObject(forKey: contactKey)
        UserDefaults.standard.removeObject(forKey: historyKey)
    }

    func send() {
        guard let contact else { return }
        let text = sendText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty else { return }
        sendText = ""
        messages.append(ChatMessage(text: text, mine: true))
        saveHistory()

        do {
            let envelope = try crypto.encryptForward(plaintext: text, recipientBoxPub: contact.boxPublicKey, me: me)
            log("→ gesendet, Ephemeral \(String(envelope.ephPub.prefix(12)))…", .info)
            transport.publish(topic: "cipherwire-demo-v1/" + contact.boxPublicKey.hexString, envelope: envelope)
        } catch {
            log("✗ " + error.localizedDescription, .fail)
        }
    }

    private func handleIncoming(_ envelope: Envelope) {
        guard let contact, envelope.senderBoxPub == contact.boxPublicKey.hexString else {
            log("✗ Nachricht von unbekannter/nicht hinzugefügter ID ignoriert", .warn)
            return
        }
        do {
            let text = try crypto.decryptForward(envelope, expectedSignPub: contact.signPublicKey, myBoxSecret: me.boxSecretKey)
            messages.append(ChatMessage(text: text, mine: false))
            saveHistory()
            log("✓ empfangen, Signatur + Verschlüsselung verifiziert", .ok)
        } catch {
            log("✗ Empfangene Nachricht verworfen: " + error.localizedDescription, .fail)
        }
    }

    // MARK: - Gegenproben

    func runTamperTest() {
        let targetPub = contact?.boxPublicKey ?? me.boxPublicKey
        guard let envelope = try? crypto.encryptForward(plaintext: "Testnachricht für Manipulationsprobe", recipientBoxPub: targetPub, me: me),
              var cipherBytes = Data(hexString: envelope.ciphertext) else { return }
        cipherBytes[0] ^= 0xFF
        let tampered = Envelope(
            senderBoxPub: envelope.senderBoxPub, senderSignPub: envelope.senderSignPub,
            ephPub: envelope.ephPub, ephSig: envelope.ephSig,
            nonce: envelope.nonce, ciphertext: cipherBytes.hexString
        )
        log("Gegenprobe: 1 Bit im Ciphertext verändert …", .info)
        do {
            _ = try crypto.decryptForward(tampered, expectedSignPub: me.signPublicKey, myBoxSecret: me.boxSecretKey)
            log("✗ Fehler: Manipulation hätte erkannt werden müssen!", .fail)
        } catch {
            log("✓ korrekt abgelehnt: " + error.localizedDescription, .ok)
        }
    }

    func runForgedSignatureTest() {
        guard let attacker = try? crypto.generateIdentity(),
              let forged = try? crypto.encryptForward(plaintext: "gefälscht", recipientBoxPub: me.boxPublicKey, me: attacker)
        else { return }
        // Envelope gibt vor "ich" zu sein, ist aber mit dem
        // Signaturschlüssel eines Angreifers signiert.
        let spoofed = Envelope(
            senderBoxPub: me.boxPublicKey.hexString, senderSignPub: me.signPublicKey.hexString,
            ephPub: forged.ephPub, ephSig: forged.ephSig,
            nonce: forged.nonce, ciphertext: forged.ciphertext
        )
        log("Gegenprobe: gefälschte Ephemeral-Signatur (falscher Signaturschlüssel) …", .info)
        do {
            _ = try crypto.decryptForward(spoofed, expectedSignPub: me.signPublicKey, myBoxSecret: me.boxSecretKey)
            log("✗ Fehler: gefälschte Signatur hätte erkannt werden müssen!", .fail)
        } catch {
            log("✓ korrekt abgelehnt: " + error.localizedDescription, .ok)
        }
    }

    func clearLog() { techLog.removeAll() }

    // MARK: - Persistenz (Klartext-UserDefaults, siehe IdentityStore.swift)

    private func loadContact() {
        guard let s = UserDefaults.standard.string(forKey: contactKey),
              let parsed = crypto.parseContactIdString(s) else { return }
        contact = parsed
    }

    private func loadHistory() {
        guard let data = UserDefaults.standard.data(forKey: historyKey),
              let decoded = try? JSONDecoder().decode([StoredMessage].self, from: data) else { return }
        messages = decoded.map { ChatMessage(text: $0.text, mine: $0.mine) }
    }

    private func saveHistory() {
        let stored = messages.map { StoredMessage(text: $0.text, mine: $0.mine) }
        guard let data = try? JSONEncoder().encode(stored) else { return }
        UserDefaults.standard.set(data, forKey: historyKey)
    }
}

private struct StoredMessage: Codable {
    let text: String
    let mine: Bool
}

File: docs/secure-messaging-architecture/client/android/SecureMessengerDemo/app/src/main/java/com/example/securemessenger/demo/AppModel.kt
/*
 * AppModel.kt
 *
 * Verbindet Identität, Kontakt, Transport und Krypto zu dem in
 * ../../../../../../../../PROTOCOL.md beschriebenen Ablauf. Hält den
 * Zustand als Compose State; alle Callbacks von Transport (Paho-Threads)
 * springen über Handler(mainLooper) auf den Haupt-Thread zurück, bevor sie
 * State verändern -- Compose State darf nur vom Haupt-Thread aus
 * geschrieben werden.
 */

package com.example.securemessenger.demo

import android.content.Context
import android.os.Handler
import android.os.Looper
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import org.json.JSONArray
import org.json.JSONObject

data class ChatMessage(val text: String, val mine: Boolean)
data class LogEntry(val text: String, val kind: LogKind)

private const val PREFS = "cipherwire_app_prefs"
private const val KEY_CONTACT = "contact_v1"
private const val KEY_HISTORY = "history_v1"

class AppModel(private val context: Context) {

    val crypto = DemoCrypto()
    val transport = Transport()
    val me: Identity = IdentityStore.loadOrCreate(context, crypto)

    var contact by mutableStateOf<ContactIdentity?>(null)
        private set
    var contactInputText by mutableStateOf("")
    var sendText by mutableStateOf("")
    val messages = mutableStateListOf<ChatMessage>()
    val techLog = mutableStateListOf<LogEntry>()

    var connectionStatus by mutableStateOf(ConnectionStatus.CONNECTING)
        private set
    var connectionStatusText by mutableStateOf("Verbinde…")
        private set

    private val mainHandler = Handler(Looper.getMainLooper())
    private fun onMain(block: () -> Unit) = mainHandler.post(block)

    init {
        loadContact()
        loadHistory()

        transport.onStatus = { status, text ->
            onMain { connectionStatus = status; connectionStatusText = text }
        }
        transport.onLog = { text, kind -> onMain { log(text, kind) } }
        transport.onEnvelope = { envelope -> onMain { handleIncoming(envelope) } }

        transport.connect(me.boxPublicKey.toHex())
        log("Meine ID: " + me.idString, LogKind.INFO)
    }

    fun log(text: String, kind: LogKind) { techLog.add(LogEntry(text, kind)) }
    fun clearLog() { techLog.clear() }

    fun addContact() {
        val parsed = crypto.parseContactIdString(contactInputText)
        if (parsed == null) {
            log("✗ Ungültiges ID-Format (erwartet: 64 Hex . 64 Hex)", LogKind.FAIL)
            return
        }
        contact = parsed
        messages.clear()
        contactInputText = ""
        saveContact(parsed)
        saveHistory()
    }

    fun removeContact() {
        contact = null
        messages.clear()
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit()
            .remove(KEY_CONTACT).remove(KEY_HISTORY).apply()
    }

    fun send() {
        val c = contact ?: return
        val text = sendText.trim()
        if (text.isEmpty()) return
        sendText = ""
        messages.add(ChatMessage(text, mine = true))
        saveHistory()

        try {
            val envelope = crypto.encryptForward(text, c.boxPublicKey, me)
            log("→ gesendet, Ephemeral ${envelope.ephPub.take(12)}…", LogKind.INFO)
            transport.publish("cipherwire-demo-v1/" + c.boxPublicKey.toHex(), envelope)
        } catch (e: Exception) {
            log("✗ ${e.message}", LogKind.FAIL)
        }
    }

    private fun handleIncoming(envelope: Envelope) {
        val c = contact
        if (c == null || envelope.senderBoxPub != c.boxPublicKey.toHex()) {
            log("✗ Nachricht von unbekannter/nicht hinzugefügter ID ignoriert", LogKind.WARN)
            return
        }
        try {
            val text = crypto.decryptForward(envelope, c.signPublicKey, me.boxSecretKey)
            messages.add(ChatMessage(text, mine = false))
            saveHistory()
            log("✓ empfangen, Signatur + Verschlüsselung verifiziert", LogKind.OK)
        } catch (e: Exception) {
            log("✗ Empfangene Nachricht verworfen: ${e.message}", LogKind.FAIL)
        }
    }

    // ---------- Gegenproben ----------

    fun runTamperTest() {
        val targetPub = contact?.boxPublicKey ?: me.boxPublicKey
        val envelope = try { crypto.encryptForward("Testnachricht für Manipulationsprobe", targetPub, me) } catch (e: Exception) { return }
        val tamperedBytes = envelope.ciphertext.fromHex()
        tamperedBytes[0] = (tamperedBytes[0].toInt() xor 0xFF).toByte()
        val tampered = envelope.copy(ciphertext = tamperedBytes.toHex())
        log("Gegenprobe: 1 Bit im Ciphertext verändert …", LogKind.INFO)
        try {
            crypto.decryptForward(tampered, me.signPublicKey, me.boxSecretKey)
            log("✗ Fehler: Manipulation hätte erkannt werden müssen!", LogKind.FAIL)
        } catch (e: Exception) {
            log("✓ korrekt abgelehnt: ${e.message}", LogKind.OK)
        }
    }

    fun runForgedSignatureTest() {
        val attacker = crypto.generateIdentity()
        val forged = try { crypto.encryptForward("gefälscht", me.boxPublicKey, attacker) } catch (e: Exception) { return }
        // Envelope gibt vor "ich" zu sein, ist aber mit dem
        // Signaturschlüssel eines Angreifers signiert.
        val spoofed = forged.copy(senderBoxPub = me.boxPublicKey.toHex(), senderSignPub = me.signPublicKey.toHex())
        log("Gegenprobe: gefälschte Ephemeral-Signatur (falscher Signaturschlüssel) …", LogKind.INFO)
        try {
            crypto.decryptForward(spoofed, me.signPublicKey, me.boxSecretKey)
            log("✗ Fehler: gefälschte Signatur hätte erkannt werden müssen!", LogKind.FAIL)
        } catch (e: Exception) {
            log("✓ korrekt abgelehnt: ${e.message}", LogKind.OK)
        }
    }

    // ---------- Persistenz (Klartext-SharedPreferences) ----------

    private fun loadContact() {
        val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        val s = prefs.getString(KEY_CONTACT, null) ?: return
        contact = crypto.parseContactIdString(s)
    }

    private fun saveContact(c: ContactIdentity) {
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit()
            .putString(KEY_CONTACT, c.boxPublicKey.toHex() + "." + c.signPublicKey.toHex())
            .apply()
    }

    private fun loadHistory() {
        val prefs = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        val s = prefs.getString(KEY_HISTORY, null) ?: return
        try {
            val arr = JSONArray(s)
            for (i in 0 until arr.length()) {
                val obj = arr.getJSONObject(i)
                messages.add(ChatMessage(obj.getString("text"), obj.getBoolean("mine")))
            }
        } catch (e: Exception) { /* korrupte/alte Historie ignorieren */ }
    }

    private fun saveHistory() {
        val arr = JSONArray()
        messages.forEach { m ->
            arr.put(JSONObject().apply { put("text", m.text); put("mine", m.mine) })
        }
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit()
            .putString(KEY_HISTORY, arr.toString())
            .apply()
    }
}
